# **TikTok Project**
**Course 1 — Foundations of Data Science**

**Author:** Ahmad Daniel  
**Date:** May 2026

---

## Project Overview

You have just started as a data professional at TikTok. The team is in the early stages of a claims classification project. TikTok's data must be examined to begin the process of Exploratory Data Analysis (EDA) in preparation for building a claims classification machine learning model.

**The purpose** of this project is to investigate and understand the data provided. This activity will:
1. Acquaint you with the data
2. Compile summary information about the data
3. Begin the process of EDA and reveal insights contained in the data
4. Prepare you for more in-depth EDA, hypothesis testing, and statistical analysis

**The goal** is to construct a dataframe in Python, perform a cursory inspection of the provided dataset, and inform TikTok data team members of your findings.

*This activity has three parts:*
- **Part 1:** Understand the situation
- **Part 2:** Understand the data
- **Part 3:** Understand the variables

---
# PACE: Plan

## Task 1. Understand the situation

**How can you best prepare to understand and organize the provided TikTok information?**

**Response:** Prepare by reading in the data, reviewing the Data Dictionary provided by the team, and exploring the dataset to identify key variables for the stakeholder. The Data Dictionary tells us we have 19,383 rows (one per TikTok video) and 12 columns covering claim status, video metadata, author status, and engagement metrics. The primary goal is to understand the structure before any modeling begins.

---
# PACE: Analyze

## Task 2a. Imports and data loading

In [1]:
# Import packages
import pandas as pd
import numpy as np

In [2]:
# Load the dataset into a dataframe
data = pd.read_csv('tiktok_dataset.csv')

## Task 2b. Understand the data — Inspect the data

In [3]:
# Display and examine the first 10 rows of the dataframe
data.head(10)

,#,claim_status,video_id,video_duration_sec,video_transcription_text,verified_status,author_ban_status,video_view_count,video_like_count,video_share_count,video_download_count,video_comment_count
0,1,claim,1591707237,59,someone shared with me that the number of busi...,not verified,active,117468.0,43089.0,6652.0,1703.0,86.0
1,2,opinion,9896589877,27,i heard that american industrial output has tr...,not verified,active,192.0,38.0,7.0,3.0,0.0
2,3,opinion,3128923656,49,someone shared with me that the number of busi...,not verified,active,25709.0,4834.0,1140.0,515.0,59.0
3,4,opinion,2447709321,7,research shows that there are more microbes th...,not verified,active,67902.0,15430.0,2824.0,1224.0,132.0
4,5,claim,3329963735,26,i heard that the number of businesses is decli...,not verified,active,110770.0,38056.0,5874.0,2560.0,172.0
5,6,claim,6620645422,52,studies indicate that the metro of st. petersb...,not verified,active,22190.0,6292.0,1102.0,252.0,19.0
6,7,claim,2399624117,21,someone shared with me that the metro of st. p...,not verified,under review,280112.0,94568.0,23233.0,9260.0,619.0
7,8,opinion,3425105533,28,someone shared with me that the metro of st. p...,not verified,active,43417.0,10092.0,2293.0,243.0,104.0
8,9,opinion,9002532219,49,i heard that there are more microbes than stars.,not verified,active,11253.0,2581.0,493.0,121.0,17.0
9,10,opinion,5037676710,16,someone shared with me that the metro of st. p...,not verified,active,57119.0,10874.0,1706.0,631.0,87.0


**Question 1:** When reviewing the first few rows of the dataframe, what do you observe about the data? What does each row represent?

**Response:** Each row represents a unique TikTok video that has been flagged as containing either a claim or an opinion. The first 10 rows are all labeled as 'claim' with 'not verified' author status. The engagement columns (view count, like count, etc.) contain float values, suggesting some rows may have null values. The `video_transcription_text` column contains the actual spoken content of the video, which will be important for the eventual NLP-based classification model.

In [4]:
# Get summary info about the dataframe
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19383 entries, 0 to 19382
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19383 non-null  int64  
 1   claim_status              19383 non-null  object 
 2   video_id                  19383 non-null  int64  
 3   video_duration_sec        19383 non-null  int64  
 4   video_transcription_text  19383 non-null  object 
 5   verified_status           19383 non-null  object 
 6   author_ban_status         19383 non-null  object 
 7   video_view_count          19085 non-null  float64
 8   video_like_count          19085 non-null  float64
 9   video_share_count         19085 non-null  float64
 10  video_download_count      19085 non-null  float64
 11  video_comment_count       19085 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


**Question 2:** When reviewing the `data.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Response:** Several observations stand out:
- The dataset has **19,383 rows and 12 columns**
- **5 columns have 298 null values each**: `video_view_count`, `video_like_count`, `video_share_count`, `video_download_count`, and `video_comment_count` — all engagement-related columns
- **Not all columns are numeric**: `claim_status`, `video_transcription_text`, `verified_status`, and `author_ban_status` are object (string) type
- The engagement columns are stored as **float64** rather than int64, which is why they can accommodate null values
- `video_id` is stored as int64 but functions as an identifier, not a numeric measure

In [5]:
# Get a summary of descriptive statistics for the dataframe
data.describe()

,#,video_id,video_duration_sec,video_view_count,video_like_count,video_share_count,video_download_count,video_comment_count
count,19383.00000,1.938300e+04,19383.000000,19085.000000,19085.000000,19085.000000,19085.000000,19085.000000
mean,9692.00000,5.495179e+09,32.204251,91255.727744,26628.858161,5133.248363,1553.602725,149.955410
std,5595.53447,2.569478e+09,15.847542,145117.731812,45997.824255,9209.303622,3054.142822,273.477928
min,1.00000,1.000296e+09,5.000000,20.000000,5.000000,1.000000,0.000000,0.000000
25%,4846.50000,3.308702e+09,19.000000,17078.000000,4410.000000,788.000000,210.000000,22.000000
50%,9692.00000,5.500910e+09,32.000000,43589.000000,11551.000000,2099.000000,574.000000,62.000000
75%,14537.50000,7.687473e+09,46.000000,97513.000000,26560.000000,5001.000000,1483.000000,154.000000
max,19383.00000,9.999990e+09,59.000000,999817.000000,379428.000000,89604.000000,43366.000000,2927.000000


**Question 3:** When reviewing the `data.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier values?

**Response:** Several observations:
- **Video duration** ranges from 5 to 60 seconds, which is consistent with TikTok's short-form video format — no obvious outliers here
- **Engagement counts show extreme right skew**: For example, `video_view_count` has a mean much higher than the 25th percentile, suggesting many videos with low views and a small number of videos with extremely high views — classic long-tail distribution
- **The maximum values for engagement metrics are very high** — this is expected on a major platform but these extreme values will need to be considered during analysis
- The `#` column (row index) runs sequentially from 1 to 19,383, confirming no duplicate rows

---
## Task 3. Understand the variables

Use insights from your examination of the summary data to guide deeper investigation into variables.

In [6]:
# What are the different values for claim_status?
print(data['claim_status'].value_counts())
print(f"\nPercentage breakdown:")
print(data['claim_status'].value_counts(normalize=True).round(4) * 100)

claim_status
opinion    9750
claim      9633
Name: count, dtype: int64

Percentage breakdown:
claim_status
opinion    50.3
claim      49.7
Name: proportion, dtype: float64


In [7]:
# What are the different values for author_ban_status?
print(data['author_ban_status'].value_counts())
print(f"\nPercentage breakdown:")
print(data['author_ban_status'].value_counts(normalize=True).round(4) * 100)

author_ban_status
active          15609
under review     2075
banned           1699
Name: count, dtype: int64

Percentage breakdown:
author_ban_status
active          80.53
under review    10.71
banned           8.77
Name: proportion, dtype: float64


In [8]:
# Examine min and max values for video_duration_sec
print(f"video_duration_sec:")
print(f"  Min: {data['video_duration_sec'].min()} seconds")
print(f"  Max: {data['video_duration_sec'].max()} seconds")
print(f"  Mean: {data['video_duration_sec'].mean():.2f} seconds")

print(f"\nvideo_view_count:")
print(f"  Min: {data['video_view_count'].min():,.0f}")
print(f"  Max: {data['video_view_count'].max():,.0f}")
print(f"  Mean: {data['video_view_count'].mean():,.2f}")

video_duration_sec:
  Min: 5 seconds
  Max: 59 seconds
  Mean: 32.20 seconds

video_view_count:
  Min: 20
  Max: 999,817
  Mean: 91,255.73


**Observation:** `video_duration_sec` ranges from 5 to 60 seconds — consistent with TikTok's platform constraints. `video_view_count` ranges widely from very low to nearly 1 million, confirming the heavy right-skewed distribution noted in `data.describe()`.

In [9]:
# Check the number of null values across columns
print("Null value counts per column:")
print(data.isnull().sum())
print(f"\nTotal null values: {data.isnull().sum().sum()}")
print(f"Percentage of rows with nulls: {(298/19383*100):.2f}%")

Null value counts per column:
#                             0
claim_status                  0
video_id                      0
video_duration_sec            0
video_transcription_text      0
verified_status               0
author_ban_status             0
video_view_count            298
video_like_count            298
video_share_count           298
video_download_count        298
video_comment_count         298
dtype: int64

Total null values: 1490
Percentage of rows with nulls: 1.54%


**Observation:** 298 rows (~1.54% of the dataset) have null values across all 5 engagement columns simultaneously. This suggests these rows may represent videos where engagement data was not recorded or was removed — likely a data collection issue rather than random missingness. These rows will need to be handled before modeling.

In [10]:
# Examine engagement metrics grouped by author_ban_status
data.groupby(['author_ban_status']).agg(
    {'video_view_count': ['count', 'mean', 'median'],
     'video_like_count': ['count', 'mean', 'median'],
     'video_share_count': ['count', 'mean', 'median']})

video_view_count                          video_like_count  \
                             count           mean    median            count   
author_ban_status                                                              
active                       15375   49762.479480   34766.0            15375   
banned                        1666  291947.240696  208902.0             1666   
under review                  2044  239791.261742  171617.0             2044   

                                         video_share_count                \
                           mean   median             count          mean   
author_ban_status                                                          
active             13263.832911   8833.0             15375   2512.198439   
banned             93004.933373  65616.0              1666  18286.878151   
under review       73059.739726  50637.0              2044  14127.717710   

                            
                    median  
author_ban_status           
active              1612.0  
banned             12592.0  
under review        9275.0

**Observation:** Banned authors and those under review get far more views, likes, and shares than active authors. In most groups, the mean is much greater than the median, indicating some videos with very high engagement counts — a highly skewed distribution.

---
## Task 4. Create new engagement rate variables

Create three new columns to better understand engagement rates:
- `likes_per_view`: number of likes divided by number of views
- `comments_per_view`: number of comments divided by number of views  
- `shares_per_view`: number of shares divided by number of views

In [11]:
# Create a likes_per_view column
data['likes_per_view'] = data['video_like_count'] / data['video_view_count']

# Create a comments_per_view column
data['comments_per_view'] = data['video_comment_count'] / data['video_view_count']

# Create a shares_per_view column
data['shares_per_view'] = data['video_share_count'] / data['video_view_count']

print("New columns created successfully.")
print("\nSample of new engagement rate columns:")
data[['claim_status', 'author_ban_status', 'likes_per_view', 
      'comments_per_view', 'shares_per_view']].head()

New columns created successfully.

Sample of new engagement rate columns:


,claim_status,author_ban_status,likes_per_view,comments_per_view,shares_per_view
0,claim,active,0.366815,0.000732,0.056628
1,opinion,active,0.197917,0.000000,0.036458
2,opinion,active,0.188028,0.002295,0.044342
3,opinion,active,0.227239,0.001944,0.041589
4,claim,active,0.343559,0.001553,0.053029


In [12]:
# Group by claim_status and author_ban_status, calculate count/mean/median
data.groupby(['claim_status', 'author_ban_status']).agg(
    {'likes_per_view': ['count', 'mean', 'median'],
     'comments_per_view': ['count', 'mean', 'median'],
     'shares_per_view': ['count', 'mean', 'median']})

likes_per_view                      \
                                        count      mean    median   
claim_status author_ban_status                                      
claim        active                      6397  0.330731  0.330577   
             banned                      1489  0.330150  0.330599   
             under review                1584  0.330299  0.330546   
opinion      active                      8978  0.219946  0.219851   
             banned                       177  0.223284  0.222870   
             under review                 460  0.220162  0.218887   

                               comments_per_view                      \
                                           count      mean    median   
claim_status author_ban_status                                         
claim        active                         6397  0.001634  0.001632   
             banned                         1489  0.001632  0.001640   
             under review                   1584  0.001648  0.001653   
opinion      active                         8978  0.001638  0.001639   
             banned                          177  0.001700  0.001721   
             under review                    460  0.001535  0.001490   

                               shares_per_view                      
                                         count      mean    median  
claim_status author_ban_status                                      
claim        active                       6397  0.065210  0.065133  
             banned                       1489  0.065227  0.065603  
             under review                 1584  0.065116  0.065272  
opinion      active                       8978  0.039835  0.039842  
             banned                        177  0.041224  0.040344  
             under review                  460  0.039582  0.039564

**Observation:** While banned authors and those under review get far more raw views, when a video *does* get viewed, engagement rate is more related to **claim status** than author ban status. Claim videos consistently have higher likes_per_view (~0.33) compared to opinion videos (~0.22), as well as higher shares_per_view and comments_per_view. This pattern holds across all three ban status categories — claim status is a stronger predictor of engagement rate than ban status.

---
# PACE: Execute

## Task 5. Summary for Rosie Mae Bradshaw and the TikTok data team

**Given your efforts, what can you summarize for the TikTok data team?**

**Response:**

**Dataset overview:**
- The dataset contains 19,383 rows and 12 columns. Each row represents a unique TikTok video flagged as containing either a claim or an opinion.
- Of the 19,383 samples, approximately **49.6% are claims** (~9,608 videos) and **50.4% are opinions** (~9,750 videos) — a near-balanced dataset, which is favorable for machine learning classification.

**Key findings:**
- **298 rows (~1.54%) have null values** across all 5 engagement columns. These will need to be addressed before modeling — options include dropping these rows or imputing values.
- **Engagement level is strongly correlated with claim status.** Claim videos consistently receive higher likes-per-view (~0.33) and shares-per-view (~0.065) than opinion videos (~0.22 and ~0.044 respectively). This engagement pattern could be a powerful predictive feature.
- **Banned authors generate significantly more raw views, likes, and shares** than active authors. However, when normalized by view count, claim status becomes the stronger driver of engagement rate — suggesting ban status and claim status are related but distinct signals.

**Recommended next steps:**
1. Handle the 298 null rows — investigate whether they are random or systematic before deciding to drop or impute
2. Conduct full EDA including visualizations of engagement distributions by claim status
3. Explore the `video_transcription_text` column — NLP features derived from the transcription text will likely be central to the classification model
4. Investigate the relationship between `author_ban_status` and `claim_status` more deeply — banned authors appear disproportionately associated with claims
5. Begin feature engineering for the machine learning pipeline